### PR TITLE
Input/Output Feature Schema Refactor

### PR DESCRIPTION
**Pull request by connor-mccorm**
Adds Marshmallow schema support for all input and output features and their corresponding preprocessing, encoders, and decoders.

### CODE REVIEW

**Comment by connor-mccorm** [comment]
Most tests are passing, there is one case where I need to add a regex pattern optional argument for the string input type to handle custom fill value restrictions, but once I get that done, we should be good to go.

File: ludwig/schema/preprocessing.py
Comment (martindavis) [inline ludwig/schema/preprocessing.py:501]: Hey @connor-mccorm , just a callout here that this is a bit of a beast on the Client. All 4 of these settings have to be set together right? If so, I'm thinking a grouped set of inputs that are all required make the most sense. However, given the way this is defined, we don't have an enum to pull from for `type` and we have to infer that `num_filter_bands` is an integer while `window_*_in_s` are floats.

I bring that all up to say: we have a similar situation with the `missing_value_strategy` of `fill_with_const` and `fill_value`. In those scenarios, we chose to select a reasonable default and have the user specify each setting separately.

I'm curious, what makes `audio_feature` different that we force the customer to specify a Dict instead of individual fields with reasonable defaults?

File: ludwig/schema/preprocessing.py
Comment (martindavis) [inline ludwig/schema/preprocessing.py:476]: Since we're including `fill_with_const` in the enum, do we need to include a `fill_value` option as well?

File: ludwig/schema/preprocessing.py
Comment (martindavis) [inline ludwig/schema/preprocessing.py:372]: Same comment as the audio field: since we're including `fill_with_const`, do we need to include a `fill_value` options as well?

File: ludwig/schema/preprocessing.py
Comment (martindavis) [inline ludwig/schema/preprocessing.py:551]: Editing comment: I'm not sure what a reasonable default is for Time Series. Glad to pair up on possible solutions!

File: ludwig/schema/preprocessing.py
Comment (martindavis) [inline ludwig/schema/preprocessing.py:656]: Hey @connor-mccorm  , this caused a bit of weirdness on the Client. The solution that @hungcs and I came up with for now is to make this field not required. Since Date Time has a standard ISO format, I'm curious if we could come up with a reasonable default that both Python/Ludwig and JavaScript's `Date()` library can interpret. Should we try to pair on that?

File: ludwig/schema/features/text_feature.py
Comment (martindavis) [inline ludwig/schema/features/text_feature.py:15]: Seeing a list this long made me think: is there a way to pull this list from code? I think one thing I get a little concerned about is that it will be hard to maintain this list and keep it in sync with the code.

**Review by martindavis — APPROVED** [review]
Hey @connor-mccorm , overall excellent job making the Schema docs very clean and clear in Ludwig! I've got some questions, but I don't see anything blocking this from my perspective.

File: ludwig/schema/preprocessing.py
Comment (connor-mccorm) [reply ludwig/schema/preprocessing.py:372]: See above:)

File: ludwig/schema/preprocessing.py
Comment (connor-mccorm) [reply ludwig/schema/preprocessing.py:551]: I think this is a tricky one to set since a timeseries column could come in many forms. I will defer to the #ludwig channel to see if people have ideas for a better default.

File: ludwig/schema/features/text_feature.py
Comment (ksbrar) [reply ludwig/schema/features/text_feature.py:15]: @connor-mccorm I wonder if you can use `encoder_registry` here? (tho this is just from a quick glance)

File: ludwig/schema/preprocessing.py
Comment (ksbrar) [reply ludwig/schema/preprocessing.py:656]: Not sure about what the default here should be, but to have something better than an unvalidated string...

@connor-mccorm what we could do here is define a custom marshmallow field specifically for datetime. And we would override `_json_type_mapping` as we do in the other complex fields, and in particular set the [`pattern`](https://json-schema.org/understanding-json-schema/reference/regular_expressions.html) attribute to this regex pattern: `^([\+-]?\d{4}(?!\d{2}\b))((-?)((0[1-9]|1[0-2])(\3([12]\d|0[1-9]|3[01]))?|W([0-4]\d|5[0-2])(-?[1-7])?|(00[1-9]|0[1-9]\d|[12]\d{2}|3([0-5]\d|6[1-6])))([T\s]((([01]\d|2[0-3])((:?)[0-5]\d)?|24\:?00)([\.,]\d+(?!:))?)?(\17[0-5]\d([\.,]\d+)?)?([zZ]|([\+-])([01]\d|2[0-3]):?([0-5]\d)?)?)?)?$` (ref: https://www.myintervals.com/blog/2009/05/20/iso-8601-date-validation-that-doesnt-suck/). Don't necessarily have to do this in this PR though.

@martindavis both RJSF and `Date` should interpret this pattern correctly.

Comment (connor-mccorm) [reply ludwig/schema/preprocessing.py:656]: Ya I think that a pattern for datetime format is a good call. I will add this in.

File: ludwig/schema/features/text_feature.py
Comment (justinxzhao) [reply ludwig/schema/features/text_feature.py:15]: You can use `list(encoder_registry["text"].keys())`

File: ludwig/schema/utils.py
Comment (connor-mccorm) [inline ludwig/schema/utils.py:91]: @ksbrar for datetime format validation, what do you think about just using the regex pattern validation argument in the `String()` marshmallow field that I added here instead of a new custom marshmallow field?

Comment (ksbrar) [reply ludwig/schema/utils.py:91]: good call! yes this is better, I just double checked that the pattern is plumbed through to the JSON schema automatically so nothing else is needed

File: ludwig/schema/preprocessing.py
Comment (connor-mccorm) [reply ludwig/schema/preprocessing.py:656]: Talked with @martindavis and decided that a pattern validation for another pattern validation isn't the best idea. Additionally, since the there's not an undeniably most common datetime format, without format inference, it is best to keep the default as None and just require the field when a datetime feature is selected. A future fix for this would be implementing datetime format inference for datetime format features.

File: ludwig/schema/utils.py
Comment (tgaddair) [inline ludwig/schema/utils.py:3]: Nit: python style would be `TList` instead of `tList`.

**Review by tgaddair — COMMENTED** [review]
Is the plan to wire this into the schema validation logic in a follow-up?

**Comment by tgaddair** [comment]
Sounds good @connor-mccorm !

File: ludwig/features/audio_feature.py
Comment (justinxzhao) [inline ludwig/features/audio_feature.py:436]: nit: These could be constants.

File: ludwig/schema/encoders.py
Comment (justinxzhao) [inline ludwig/schema/encoders.py:57]: Could these options be replaced with `initializer_registry.keys()`

File: ludwig/schema/encoders.py
Comment (justinxzhao) [inline ludwig/schema/encoders.py:116]: Could this be replaced with `activations.keys()` (perhaps renamed to `activations_registry`) from ludwig.utils.torch_utils?

File: ludwig/schema/preprocessing.py
Comment (justinxzhao) [inline ludwig/schema/preprocessing.py:51]: Could we restrict the options to `tokenizer_registry.keys()`?

File: ludwig/schema/preprocessing.py
Comment (justinxzhao) [inline ludwig/schema/preprocessing.py:116]: What is the difference between `fill_value` and `computed_fill_value`?

It looks like `computed_fill_value` might be an internal-only parameter that perhaps we want to leave out of the schema.

File: ludwig/schema/preprocessing.py
Comment (justinxzhao) [inline ludwig/schema/preprocessing.py:81]:
```suggestion
        description="The string used as the padding symbol for sequence features. Ignored for features using huggingface encoders, which have their own vocabulary.",
```

File: ludwig/schema/preprocessing.py
Comment (justinxzhao) [inline ludwig/schema/preprocessing.py:88]:
```suggestion
        description="The string used as the unknown symbol for sequence features. Ignored for features using huggingface encoders, which have their own vocabulary.",
```

File: ludwig/schema/preprocessing.py
Comment (justinxzhao) [inline ludwig/schema/preprocessing.py:348]: Consider factoring out a `UNIVERSAL_MISSING_VALUE_STRATEGIES` which has the missing value strategies that are shared across all features.

File: ludwig/schema/preprocessing.py
Comment (justinxzhao) [inline ludwig/schema/preprocessing.py:568]: Restrict this to `tokenizer_registry.keys()`

File: tests/ludwig/marshmallow/test_fields_preprocessing.py
Comment (justinxzhao) [inline tests/ludwig/marshmallow/test_fields_preprocessing.py:12]: Let's stick with snake case that we've been using for all of tests so far.

File: ludwig/schema/encoders.py
Comment (connor-mccorm) [reply ludwig/schema/encoders.py:57]: Yes good call. These files will actually be used in a followup PR for the encoder schema refactor since the current implementation is not actually using this file. I will include the registry in the next PR however.

File: ludwig/schema/encoders.py
Comment (connor-mccorm) [reply ludwig/schema/encoders.py:116]: Same as above, will include in the encoder schema refactor.

File: ludwig/schema/preprocessing.py
Comment (connor-mccorm) [reply ludwig/schema/preprocessing.py:51]: Same as above, will include in the encoder schema refactor.

File: ludwig/schema/preprocessing.py
Comment (connor-mccorm) [reply ludwig/schema/preprocessing.py:116]: I believe it is just an internal parameter. Specifically, I think it is the fill_value that we calculate based on the data internally. However, throughout the original schema implementation and all of the testing utils, they have `computed_fill_value` as a member of the schema so I figured I would stay consistent and keep it in. I think the best thing to do hear would be to ask #ludwig channel and see why it's in the schema, and if theres not a strong reason to keep in the schema, I can remove it throughout the entire project.

File: ludwig/schema/preprocessing.py
Comment (connor-mccorm) [reply ludwig/schema/preprocessing.py:348]: Will do!

File: ludwig/schema/preprocessing.py
Comment (connor-mccorm) [reply ludwig/schema/preprocessing.py:568]: Will do!

File: tests/ludwig/marshmallow/test_fields_preprocessing.py
Comment (connor-mccorm) [reply tests/ludwig/marshmallow/test_fields_preprocessing.py:12]: Good catch, will make the change!

File: ludwig/schema/encoders.py
Comment (connor-mccorm) [reply ludwig/schema/encoders.py:57]: Going to remove this file from this PR since it is not relevant until the encoder schema PR.

File: ludwig/schema/encoders.py
Comment (connor-mccorm) [reply ludwig/schema/encoders.py:116]: Going to remove this file from this PR since it is not relevant until the encoder schema PR.

File: ludwig/schema/preprocessing.py
Comment (connor-mccorm) [reply ludwig/schema/preprocessing.py:51]: Done!

File: ludwig/schema/preprocessing.py
Comment (connor-mccorm) [reply ludwig/schema/preprocessing.py:568]: Done!

File: tests/ludwig/marshmallow/test_fields_preprocessing.py
Comment (connor-mccorm) [reply tests/ludwig/marshmallow/test_fields_preprocessing.py:12]: Done!

File: ludwig/schema/preprocessing.py
Comment (connor-mccorm) [reply ludwig/schema/preprocessing.py:348]: Done!

File: ludwig/schema/preprocessing.py
Comment (connor-mccorm) [reply ludwig/schema/preprocessing.py:88]: Done

File: ludwig/schema/preprocessing.py
Comment (connor-mccorm) [reply ludwig/schema/preprocessing.py:81]: Done

File: ludwig/schema/preprocessing.py
Comment (connor-mccorm) [reply ludwig/schema/preprocessing.py:116]: Checked with Piero and confirmed it's internal only, so removing from schema.

File: ludwig/schema/preprocessing.py
Comment (ksbrar) [inline ludwig/schema/preprocessing.py:55]: If this isn't causing bugs then its a moot point, but it looks like you are registering the preprocessing configs before they are marshmallow-fied by the dataclass decorator.

File: ludwig/schema/optimizers.py
Comment (ksbrar) [inline ludwig/schema/optimizers.py:328]: Changed for readability?

File: ludwig/schema/features/utils.py
Comment (ksbrar) [inline ludwig/schema/features/utils.py:26]: A bit confused why these two update functions are necessary. Aren't the `encoder` props on the feature schemas declared with calls to `get_encoder_classes`/`get_decoder_classes` already?

File: ludwig/schema/__init__.py
Comment (ksbrar) [inline ludwig/schema/__init__.py:34]: SO clean 🤤

File: ludwig/features/number_feature.py
Comment (ksbrar) [inline ludwig/features/number_feature.py:244]: I believe this list gets duplicated (once here, once in the preprocessing schema), which is unfortunate. Unrelated to this PR but another indication we should do a wider registry refactor at some point.

File: ludwig/schema/preprocessing.py
Comment (ksbrar) [inline ludwig/schema/preprocessing.py:652]: Nit: update this docstring to remove `oneOf` ref.

**Review by ksbrar — APPROVED** [review]
Awesome work!!

Some questions and nits but LGTM.

File: ludwig/schema/preprocessing.py
Comment (justinxzhao) [reply ludwig/schema/preprocessing.py:128]: This is already a constant: `strings_utils.UNKNOWN_SYMBOL`

File: ludwig/schema/utils.py
Comment (connor-mccorm) [reply ludwig/schema/utils.py:2]: Done!

File: ludwig/schema/preprocessing.py
Comment (connor-mccorm) [reply ludwig/schema/preprocessing.py:197]: The rationale here is that this is a required field in the config, but it can't be `None`. If the user doesn't touch it, then it populates with the default, which is why it's typed as `Optional[str]` but if they do decide to touch it, they can't set the value to None, otherwise we'll get an "Invalid missing_value_strategy error". So the solution was to set it to `None` to avoid the default.

File: ludwig/schema/preprocessing.py
Comment (connor-mccorm) [reply ludwig/schema/preprocessing.py:128]: Will switch this out!

File: ludwig/schema/preprocessing.py
Comment (connor-mccorm) [reply ludwig/schema/preprocessing.py:652]: Done!

File: ludwig/schema/__init__.py
Comment (connor-mccorm) [reply ludwig/schema/__init__.py:34]: Iknorite?? It was really satisfying replacing with these functions🤠

File: ludwig/schema/features/utils.py
Comment (ksbrar) [reply ludwig/schema/features/utils.py:26]: gotcha! please add docstring or comment where it's called with this note

Comment (connor-mccorm) [reply ludwig/schema/features/utils.py:26]: Will do!

File: ludwig/schema/preprocessing.py
Comment (connor-mccorm) [reply ludwig/schema/preprocessing.py]: Now moving to `parameter_metadata` schema validation method with `internal_only` flag.

File: ludwig/schema/preprocessing.py
Comment (ksbrar) [reply ludwig/schema/preprocessing.py:197]: Might be misunderstanding you, but the default choice is `fill_with_const`, right? Imagining I delete `Optional` for a moment and then the functionality is: if a user types nothing, `fill_with_const` is what gets chosen; at the same time, setting `allow_none` to `False` prevents the user from typing `missing_value_strategy: null` in the config. So `missing_value_strategy` can never be `None`. That seems like what we want, right?

File: ludwig/schema/features/text_feature.py
Comment (connor-mccorm) [reply ludwig/schema/features/text_feature.py:25]: Will move to the schema generation logic since you are right, we don't really need to define this in the schema if it's consistent across all features.

Comment (connor-mccorm) [reply ludwig/schema/features/text_feature.py]: Done!

**Comment by connor-mccorm** [comment]
Update on this: waiting for https://github.com/ludwig-ai/ludwig/pull/2224 to merge this so that I can use the `internal_only` parameter metadata flag for `computed_fill_value` in the preprocessing schemas.

File: ludwig/schema/features/text_feature.py
Comment (connor-mccorm) [reply ludwig/schema/features/text_feature.py]: So after making this change, it looks like a bunch of the tests are failing now because they want to set tied to `None` which is not of type string. Was looking for a way to do that, however it got me thinking that since we want to create this schema object for the SDK that basically uses the schema definitions for each input feature type in order to create input/output features via the SDK using dot notation, it may be better to actually have these constants defined in the schema classes themselves like `name`, `type` and `tied`. So that the user can modify all these attributes when creating these objects via the SDK object. WDYT, @justinxzhao?

Comment (ksbrar) [reply ludwig/schema/features/text_feature.py]: @connor-mccorm missing some context, but would defining some tertiary base/common class for fields like this and deriving it work, e.g. "`CommonFeatureOptions`"

Comment (connor-mccorm) [reply ludwig/schema/features/text_feature.py]: Ya that's probably the move actually. I had that before but ran into an inheritance issue that I resolved by just removing the class entirely which was dumb of me. Didn't have the foresight to see that we would probably need that at the time. Will add now.

**Comment by justinxzhao** [comment]
@connor-mccorm FYI #2224 has now merged.

**Comment by connor-mccorm** [comment]
Thank you @justinxzhao, will work on merging this today!

**Pull request merged**
